src/core/ctxt.ml

```diff
@@ -2,66 +2,99 @@
+  let (_,a,_) = List.find (fun (y,_,_) -> Bindlib.eq_vars x y) ctx in a
 
-(** [add x a ctx] maps the variable [x] to the type [a] in [ctx]. *)
-let add : tvar -> term -> t -> t =
-  fun x a ctx -> (x,a)::ctx
-
-(** [unbind_ctxt ctx a b] returns the triple [(x,b,ctx')] such that [(x,b)]
-   is the unbinding of [b] and [ctx'] is the context [ctx] extended with
-   [(x,a)] if [x] occurs in [b]. *)
-let unbind ctx a b =
-  let (x,b') = Bindlib.unbind b in
-  let ctx' = if Bindlib.binder_occur b then add x a ctx else ctx in
-  (x,b',ctx')
-
-(** [pp oc ctx] prints the context [ctx] to the channel [oc]. *)
-let pp : t pp = fun oc ctx ->
-  let pp_e oc (x,a) =
-    Format.fprintf oc "%a : %a" Print.pp_tvar x Print.pp a
-  in
-  if ctx = [] then Format.pp_print_string oc "∅"
-  else List.pp pp_e ", " oc (List.rev ctx)
-
-(** [find x ctx] returns the type of [x] in the context [ctx] when it appears,
-    and raises [Not_found] otherwise. *)
-let type_of : tvar -> t -> term = fun x ctx ->
-  snd (List.find (fun (y,_) -> Bindlib.eq_vars x y) ctx)
+(** [def_of x ctx] returns the definition of [x] in the context [ctx] if it
+    appears, and [None] otherwise *)
+let rec def_of : term Bindlib.var -> ctxt -> term option = fun x ctx ->
+  match ctx with
+  | []         -> None
+  | (y,_,d)::l -> if Bindlib.eq_vars x y then d else def_of x l
 
 (** [mem x ctx] tells whether variable [x] is mapped in the context [ctx]. *)
-let mem : tvar -> t -> bool = fun x ctx ->
-  try ignore (type_of x ctx); false with Not_found -> true
+let mem : tvar -> ctxt -> bool = fun x ->
+  List.exists (fun (y,_,_) -> Bindlib.eq_vars x y)
 
 (** [to_prod ctx t] builds a product by abstracting over the context [ctx], in
-    the term [t]. *)
-let to_prod : t -> term -> term = fun ctx t ->
-  match ctx with
-  | []      -> t
-  | [(x,a)] -> Prod(a, Bindlib.unbox (Bindlib.bind_var x (lift t)))
-  | _       -> let fn t (x,a) = _Prod (lift a) (Bindlib.bind_var x t) in
-               Bindlib.unbox (List.fold_left fn (lift t) ctx)
-
-(** [of_env] builds a context from an environment. **)
-let of_env : Env.t -> t =
-  List.map (fun (_,(v,bt)) -> v,Bindlib.unbox bt)
+    the term [t]. It returns the number of products as well. *)
+let to_prod : ctxt -> term -> term * int = fun ctx t ->
+  let fn (t,c) elt =
+    match elt with
+    | (x,a,None   ) -> (_Prod (lift a) (Bindlib.bind_var x t), c + 1)
+    | (x,a,Some(u)) -> (_LLet (lift u) (lift a) (Bindlib.bind_var x t), c)
+  in
+  let t, c = List.fold_left fn (lift t, 0) ctx in
+  (Bindlib.unbox t, c)
 
-(** [make_meta ctx a] creates a metavariable of type [a],  with an environment
-    containing the variables of context [ctx]. *)
-let make_meta : t -> term -> term = fun ctx a ->
-  let m = fresh_meta (to_prod ctx a) (List.length ctx) in
-  Meta(m, Array.of_list (List.rev_map (fun (v,_) -> Vari v) ctx))
+(** [to_llet ctx t] builds one let-binding on top of [t] for each defined
+    variable in [ctx]. *)
+let rec to_llet ctx t =
```

Comment:
I don't really like that name, why is this function called this way ("Llet")?

Reply:
Because it creates `LLet`s from the context, like `to_prod` creates `Prod`s from the context.